src/js/multiPortConnector.js

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2022, Tony Atkins
+ * Copyright 2011-2020, Colin Clark
+ *
+ *  Licensed under the MIT license, see LICENSE for details.
+ */
+
+(function (fluid) {
+    "use strict";
+    var youme = fluid.registerNamespace("youme");
+
+    // Enclosing grade so that we can recreate our pool of connections.
+    fluid.defaults("youme.multiPortConnector", {
+        gradeNames: ["fluid.modelComponent"],
+
+        direction: "inputs",
+
+        model: {
+            connectionPorts: [],
+            portSpecs: [],
+            ports: "{youme.system}.model.ports",
+            open: true
+        },
+
+        components: {
+            midiSystem: {
+                type: "youme.system"
+            }
+        },
+
+        dynamicComponents: {
+            connection: {
+                type: "youme.connection",
+                sources: "{youme.multiPortConnector}.model.connectionPorts",
+                options: {
+                    model: {
+                        port: "{source}",
+                        open: "{youme.multiPortConnector}.model.open"
+                    }
+                }
+            }
+        },
+
+        modelListeners: {
+            "ports": {
```

Comment:
These might be better written using the "listen on multiple paths" syntax - https://docs.fluidproject.org/infusion/development/changeapplierapi#the-path-entry-in-a-model-listener-declaration
i.e. with a member path: ["ports", "portSpecs"] and then you can use the key for a namespace, e.g. findPorts

Reply:
Or alternatively simply as a modelRelay specification which will take care of invalidation for you,e.g.
````
modelRelay: {
    findPorts: {
        target: "connectionPorts",
        func: "youme.multiPortConnector.findPorts",
        args: ["{that}", "{that}.options.direction", "{that}.model.ports", "{that}.model.portSpecs"]
    }
}
````

Reply:
If I use the suggested code (which sounds great, BTW), I get circular reference errors, as in:

```
Error: Assertion failure - check console for more details: Error in component configuration - a circular reference was found during evaluation of path segment "model of path ["model"]": for more details, see the activity records following this message in the console, or issue fluid.setLogging(fluid.logLevel.TRACE) when running your application
```

Reply:
Sadly, the multi-path relay doesn't seem to work either, which is not too surprising, as it seems to longer be tested.  Here's [a codepen demonstrating the lack of relay](https://codepen.io/duhrer/pen/abYmzKy).

Reply:
> Sadly, the multi-path relay doesn't seem to work either, which is not too surprising, as it seems to longer be tested. Here's [a codepen demonstrating the lack of relay](https://codepen.io/duhrer/pen/abYmzKy).

It is tested here - https://github.com/fluid-project/infusion/blob/main/tests/framework-tests/core/js/DataBindingTests.js#L1923-L1927 the correct syntax is to use "path" which works in the codepen

Reply:
> If I use the suggested code (which sounds great, BTW), I get circular reference errors, as in:

Could you check in a branch for me to investigate, cheers

Reply:
That's also why I thought there were no tests, as I was searching for `paths`.

Reply:
Or maybe not, I don't see any usages of array `path` values either in the tests, but it does at least work.

Reply:
>> If I use the suggested code (which sounds great, BTW), I get circular reference errors, as in:
> Could you check in a branch for me to investigate, cheers

I found this works fine if the relay definition is written to be a proper side-effect-free function as follows:

````
    youme.multiPortConnector.findPorts = function (that, direction, ports, portSpecs) {
        // Create it as a map indexed by ID so that we can avoid duplicates.
        var connectionPortsById = {};
        fluid.each(portSpecs, function (portSpec) {
            var portsToSearch = fluid.makeArray(ports && ports[direction]);
            if (portsToSearch.length > 0) {
                var foundPorts = youme.findPorts(portsToSearch, portSpec);
                fluid.each(foundPorts, function (port) {
                    connectionPortsById[port.id] = port;
                });
            }
        });
        var connectionPorts = Object.values(connectionPortsById);
        return connectionPorts;
    };
````

Note that you can't issue explicit reads to the model from inside a relay definition, these all need to be hoisted into arguments so that invalidation can properly be checked.

Reply:
Thanks, that works.  I'll give the port connector another look shortly.